Create Sugu Sougoumarane podcast post (#5035)
sourcegraph/about: podcast/2.15.md

---
title: 'Creating the GitHub of databases, with Sugu Sougoumarane, co-founder and CTO of PlanetScale'
publishDate: 2022-01-18T10:00-07:00
tags: [podcast]
slug: sugu-sougoumarane
published: true
---

<!-- START AUDIO -->

<audio className="object-center" src="https://www.buzzsprout.com/1097978/9653528-building-technical-communities-with-swyx-head-of-developer-experience-at-temporal.mp3?download=true" controls preload="none"></audio>

<!-- END AUDIO -->

<!-- START GUESTS -->
<span>
Sugu Sougoumarane, Beyang Liu
</span>
<!-- END GUESTS -->

<!-- START SUMMARY -->

> Why is using Planetscale a mind-altering experience? In this episode, Sugu Sougoumarane, co-founder and CTO of PlanetScale, shares how one email got him a second job interview with Elon Musk, tells the story of how he became one of the elite engineers at Paypay by solving the company’s most painful process, and explains why database administrators are shifting from managing machines to managing fleets of machines. Along the way, Sougoumarane explains why so many developers have told him they’ve felt like they’ve waited their wholes lives for self-serve schema deployment.

Reviewer: ```suggestion
Why is using Planetscale a mind-altering experience? In this episode, Sugu Sougoumarane, co-founder and CTO of PlanetScale, shares how one email got him a second job interview with Elon Musk, tells the story of how he became one of the elite engineers at Paypal by solving the company’s most painful process, and explains why database administrators are shifting from managing machines to managing fleets of machines. Along the way, Sougoumarane explains why so many developers have told him they’ve felt like they’ve waited their wholes lives for self-serve schema deployment.
```

I don't know if you've read the old PayPal business model. Originally, when PayPal was founded, the way we raised money was we told them that we will allow people to send money to each other, and we will make money from the float. The fact that when people are sending money with each other means that we'll have some float money, and we'll make money off of that. So that was the business model that was proposed. You can probably find it if you search for it.

But this was engineers writing code that moved money. They were inserting rows into a transaction table, updating the user balance, doing a commit, and moving on. And obviously–guess what happens–there are bugs in the software. So at some point in time, auditors came and said, "Okay, show us your books." And so they add up the transactions, they look at the user balances, and there's a few million dollars missing. And they're like, "Where's this money?" And we're like, "Oh, that's all in the float. It's float money. It's being invested somewhere."

**Beyang Liu:**

They didn't accept that answer.

**Sugu Sougoumarane:**

> That kind of hand-waving doesn't go well in the financial industry. So, they basically asked us to account for every penny. And so somebody hacked up a tool. But because developers were writing software and there were bugs, that tool used to report discrepancies every day, or every other day. And every time it posted, somebody had to go in and investigate. And in millions of transactions, how do you find out where things went wrong? And so there was a victim engineer that was selected every once in a while the activity report reported conflicts. The guy said, "Okay, my life is gone today."

Reviewer: The next to last sentence in this paragraph is a bit confusing - is there a way to clarify it with a bit of punctuation or gentle rewording?

**Sugu Sougoumarane:**

Split the database into smaller parts. And that experiment failed, not for technical reasons, but nevertheless, the experiment failed. And we had to rollback.

**Beyang Liu:**

So all PayPal was literally on a single giant database.

**Sugu Sougoumarane:**

Yes.

**Beyang Liu:**

What was the database, just out of curiosity?

**Sugu Sougoumarane:**

Oracle.

**Beyang Liu:**

Oh, Oracle. Okay. Oracle. So one giant Oracle database instance.

**Sugu Sougoumarane:**

It was Oracle, and we had the biggest hardware money could buy.

**Beyang Liu:**

Which was what in those days? Do you remember?

**Sugu Sougoumarane:**

It was called the Sun E15K, I believe. It was the biggest machine Sun ever produced. If Sun made a new, more powerful machine, we bought it. So I think we had an E10K before that. And then after that we bought something called the E15K.

**Beyang Liu:**

It was one box? Literally one box.

**Sugu Sougoumarane:**

One box. We had two E15Ks because there was a hot standby.

**Beyang Liu:**

Okay. Got it. Yeah, for fallover.

**Sugu Sougoumarane:**

So that's how Oracle ran.

**Beyang Liu:**

Makes sense.

**Sugu Sougoumarane:**

> The only hint I can give you is, after that outage we migrated out of E15K into Fujitsu. That happened after I left, but there is a hint in there.

Reviewer: Is there a prior mention of the outage referenced here? It seems to come out of the blue

**Beyang Liu:**

So what was the problem that you were working on at YouTube that necessitated scaling out MySQL? What was that? Was it storing video data, or something else, or?

**Sugu Sougoumarane:**

Oh, no. So, video data was not stored in MySQL. Video data was stored in CDNs.

**Beyang Liu:**

That makes sense.

**Sugu Sougoumarane:**

I actually don't know. I believe we used a CDN service. In other words, I don't think we stored the data ourselves.

**Beyang Liu:**

It was just like Akamai or something like that?

**Sugu Sougoumarane:**

I think it was Akamai. Most likely Akamai because that name does ring a bell. So we stored the video data and Akamai would spread it around the world and take care of that part. But it was metadata. Metadata was growing too fast. We couldn't keep up.

**Beyang Liu:**

What does metadata encompass? Is that comments and descriptions and titles?

**Sugu Sougoumarane:**

Comments, videos, user information. Apparently, the insert rate in the video was so high that it could not keep up with the inserts. Because we could not insert fast enough.

**Beyang Liu:**

> There were so many comments that people were living on videos, or?

Reviewer: ```suggestion
There were so many comments that people were leaving on videos, or?
```

**Beyang Liu:**

Eight shards should be enough for anyone.

**Sugu Sougoumarane:**

I mean, we went so far with one shard. With eight shards, we can scale 8X. Here we are all set for the rest of our lives, is what we thought. And what was happening was that, at eight shards, the shards themselves were okay. We had not reached their capacity. But what had happened was that managing those eight shards became a huge pain for a number of reasons. And there were very frequent outages. A couple of outages a day kind of thing.

**Beyang Liu:**

Because one shard would go down?

**Sugu Sougoumarane:**

There was no peace in our organization. There was no peace. Because there was always an outage going on, and every time it was the database. So it was almost like, "Oh, there's an outage. Let's start looking at the database because that's where the problem is." And most of the time, that was the case. And so the thing that my co-founder, co-creator, and I did–actually, most of that work was done by him. What he did was, he took himself out and wrote a spreadsheet. There was no postmortem process those days.

**Beyang Liu:**

No Five Why's.

**Sugu Sougoumarane:**

> Yeah, none of that stuff. But he kind of went through that. He made a spreadsheet of every failure that we had ever faced. Every one of those outages. What caused it and what can we do to prevent it? And so he wrote–iit was a huge spreadsheet. I may even be able to dig it up again. I should it somewhere.

Reviewer: ```suggestion
Yeah, none of that stuff. But he kind of went through that. He made a spreadsheet of every failure that we had ever faced. Every one of those outages. What caused it and what can we do to prevent it? And so he wrote–it was a huge spreadsheet. I may even be able to dig it up again. I should it somewhere.
```

**Beyang Liu:**

Like he manually compiled?

**Sugu Sougoumarane:**

Yeah, he sat and wrote every outage that he had seen that he remembered. So he wrote it down and asked: what could solve it?

> And so, by the end of that spreadsheet, when we looked at that spreadsheet. The goal was to actually do something that would leap ahead of all these problems, that will not only solve the problems of today, but also future problems that we are likely to have. So when we looked at the spreadsheet, it was obvious that we had to build a new software to handle all these cases.

Reviewer: ```suggestion
And so, by the end of that spreadsheet, when we looked at that spreadsheet, the goal was to actually do something that would leap ahead of all these problems, that will not only solve the problems of today, but also future problems that we are likely to have. So when we looked at the spreadsheet, it was obvious that we had to build a new software to handle all these cases.
```

**Beyang Liu:**

It's on the front page.

**Sugu Sougoumarane:**

Which means that every hit on the front page ended up fetching 250,000 videos.

**Beyang Liu:**

Got it. Got it.

**Sugu Sougoumarane:**

So one of the things that we wrote was, well, how do we protect against this? So, the way we did this was again, we had this proxy, right? So this proxy, every time it got a query, it would check if the same query is already executing.

If it is already executing, it won't send it to MySQL, it'll just wait for that result.

**Beyang Liu:**

It's kind of batching the requests.

**Sugu Sougoumarane:**

Or consolidating.

**Beyang Liu:**

Consolidating.

**Sugu Sougoumarane:**

Yeah. So this was an expensive query. At any given point of time, only one of those can run. If the site got spammed with the same query, it won't end up spamming the database. At any given point... if that finishes another one will start, but all other thundering herd requests will wait and will just share the result. So stuff like this, we started adding. And at that time, Vitess had only VT Tablet. Its only job was to protect MySQL, the single MySQL instance, that was-

**Beyang Liu:**

> Was is VT Tablet? I'm actually unfamiliar with that.

Reviewer: ```suggestion
What is VT Tablet? I'm actually unfamiliar with that.
```

**Sugu Sougoumarane:**

So VT Tablet was the proxy that we first wrote. That was essentially the Vitess project.

**Beyang Liu:**

I see. Got it.

**Sugu Sougoumarane:**

Now, it is a project that had many other components. So when we wrote it, it was not actually meant to manage sharding. It was only meant to manage your cluster.

Later that evolved into, "Hey, is it ever possible to make the system so generic that someone other than YouTube who wanted to shard could use it." And that was the beginning of what you see as Vitess today.

**Beyang Liu:**

So, how did you convince Google, the company, that this was worth open-sourcing? I mean, my understanding was, it was originally for YouTube, the internal system specifically. Did you have to fight?

**Sugu Sougoumarane:**

> Not at all. Actually, it was extremely easy and straightforward. I believe it is still that type of company where if you want to start a project, if you're starting a project and think that this is worthy of being open-sourced, the only thing they want to make sure is that you are not exposing any company secrets that are critical to Google's business. As long as that is not the case, you got approval. Our own decision to open-source was a very selfish decision at that point. If we build this system, I don't want to have to rebuild this if I went somewhere else.

Reviewer: Check consistency on "open-source" vs "open source" - I think you've got both here and below.

**Beyang Liu:**

Yeah, of course.

**Sugu Sougoumarane:**

> And there is nothing proprietary. And so we are going to open source this. And so we decided to open source it. So Vitess, actually, I don't know how many rules of Google we broke by doing what we did, but we created a sub-version repository in our home directory, and checked in all the code into that repository.

Reviewer: ```suggestion
And there is nothing proprietary. And so we are going to open source this. And so we decided to open source it. So Vitess, actually, I don't know how many rules of Google we broke by doing what we did, but we created a Subversion repository in our home directory, and checked in all the code into that repository.
```

**Beyang Liu:**

It was all in the family.

**Sugu Sougoumarane:**

It was all in the family. So it was well-understood and Google's home directory is pretty secure that way.

**Beyang Liu:**

Yeah. That's awesome.

**Sugu Sougoumarane:**

And even when we went to the Google open source team and told them this, they were not shocked or anything. They said, "Yeah, sure." It's no different from you developing some code. So yeah, because it was backed up, it was secured, we were not going to lose that file system. It had all the security, so there was no fear of...

**Beyang Liu:**

I like that. So, give me a sense of timeline. When did Vitess as a project... when was the beginning of that, and then when was it open-sourced for the first time?

**Sugu Sougoumarane:**

So I believe we started the project in 2010. It went into production in 2011, and we opened sourced it in 2012.

**Beyang Liu:**

2012, okay, got it.

**Sugu Sougoumarane:**

2012 I think is when we first open sourced it. So initially we had named it Voltron.

**Beyang Liu:**

Voltron. Is that a Pokémon or-

**Sugu Sougoumarane:**

> It's a Japanese cartoon. I think it is something where five different pieces come together to form this big giant thing. And the idea was that Vitess is a bunch of pieces that come together to. So that model still makes sense because Vitess is a distributed system where a bunch of components come and work together. So we named it Voltron, and obviously we could not open source with that name.

Reviewer: In "And the idea was that Vitess is a bunch of pieces that come together to" should that be "too" or is it a sentence that trails off?

**Beyang Liu:**

Wow.

**Sugu Sougoumarane:**

But nobody did. We felt that for YouTube to continue to use Vitess, it had to become a mainstream project in its own right. So that was one motivation.

**Beyang Liu:**

Yeah. Keep it around.

**Sugu Sougoumarane:**

Yeah. But then there were other motivations. The other motivation was eventually what is happening is, YouTube was also changing as a company, where the newer and newer talent that was coming in, they were all product-oriented.

Where does YouTube make money? YouTube does not make money by building Vitess.

YouTube makes money by producing videos, publishing them, and stuff. That's where, naturally, all the effort was going. And at some point in time, the newer management in YouTube was starting to ask, why are we funding this project? But then this project is gaining adoption outside. Everybody is using it. We need this to be funded.

And then there is Spanner, and Google internally was saying, "We don't know what this Vitess is."

**Beyang Liu:**

Just put it all in Spanner.

**Sugu Sougoumarane:**

> Put it all in Spanner. That is what is blessed within Google. You should move to that. So I started talking to people. I talked to Joe Beda and Craig McLuckie, who were the founders of Heptio. So I started talking to him, because they kind of went through the same thing. They created Kubernetes, and ended up leaving YouTube to start Heptio as a company. And they also founded the CNCF.

Reviewer: ```suggestion
Put it all in Spanner. That is what is blessed within Google. You should move to that. So I started talking to people. I talked to Joe Beda and Craig McLuckie, who were the founders of Heptio. So I started talking to them, because they kind of went through the same thing. They created Kubernetes, and ended up leaving YouTube to start Heptio as a company. And they also founded the CNCF.
```

**Beyang Liu:**

Oh, I did not know that. Interesting.

**Sugu Sougoumarane:**

They had done all that, so we kind of chatted with them. "So Vitess is in this stage where YouTube doesn't want to spend more money on it. How did you manage Kubernetes?" That's when they said, "Vitess should join CNCF. It's the perfect fit." They kind of encouraged us to do it. So I went back and proposed to YouTube that we should donate Vitess to CNCF. YouTube liked the idea. Said it's perfect. And the kind of understanding was that after that I will go and start a company on Vitess. That's kind of how this all happened. So this happened during 2017. And in 2018, we started the company.

**Beyang Liu:**

So, the company is quite young, even though Vitess has been around.

**Sugu Sougoumarane:**

> Yes. Yes. That was actually what the community was missing. Because Vitess is not not a tooling that you try today, you don't like it, you throw it out tomorrow. It stores your core data, your core infrastructure.

Reviewer: ```suggestion
Yes. Yes. That was actually what the community was missing. Because Vitess is not a tooling that you try today, you don't like it, you throw it out tomorrow. It stores your core data, your core infrastructure.
```

**Beyang Liu:**

Yeah. I've been there.

**Sugu Sougoumarane:**

One small fat finger, and you have an enormous outage of epic proportions. So that's very scary. You cannot just say, "Here's a schema change. Boom done." This danger exists, right? How do companies solve the problem? They install gatekeepers. People to review your change.

**Beyang Liu:**

Only certain people can directly access the database and you have to go through them.

**Sugu Sougoumarane:**

Only they can directly access data, and they will review your code. And I have a schema that is ready to be deployed, but this DBA is on vacation, now I have to wait three days for them to come back. Or, they are busy fixing, taking backups, so they are not available. Because a DBA's job is not to wait. That's not their priority to review your schema change. So the entire process becomes a huge headache. And so Sam, who is our chief product officer, he kind of developed that system within GitHub. He came, he said, "I know what the developers want, and this is what we're doing..." So this whole thing was his idea.

> And you can see how the developer community is responding to these features. "I've been waiting for this all my life," is what the developers are saying, which is self-serve schema safe schema deployments.

Reviewer: Should "schema safe" be "schema-safe"?

**Sugu Sougoumarane:**

With none of the hassle. And also, the way we deploy the schema is underlying... I don't know if you've heard of gh-ost, which is the MySQL schema deployer.

**Beyang Liu:**

No.

**Sugu Sougoumarane:**

Basically, it allows you to deploy schemas without downtime.

So if you have a billion rows in your table and you want to add a column, MySQL is going to lock that table, and you're not going to be able to write to it. And gh-ost actually allows you to do the deployment without locking that table. It's called gh-ost, which stands for GitHub Online Schema Tool. But it's also gh-ost, because it creates a ghost table for you.

**Beyang Liu:**

Oh, I see. That's the trick. It just copies stuff over.

**Sugu Sougoumarane:**

And basically it forms an atomic switch. And it turns out that Vitess has the same technology called V-replication as a core primitive that you can use.

**Beyang Liu:**

That's convenient.

**Sugu Sougoumarane:**

And the author of gh-ost coincidentally happens to work for PlanetScale, and coincidentally decides to implement that same feature in Vitess. So the same offline schema deployment now is available through Vitess.

**Beyang Liu:**

> I feel like every organization or company that operates a database, they over time accumulate a bunch of scripts that automate certain tasks, whether it's schema deploy, or pulling certain pieces of information. It sounds like you're just productizing all of that and turning it into an out-of-the-box, non hacky, non-duct-tape solution.

Reviewer: ```suggestion
I feel like every organization or company that operates a database, they over time accumulate a bunch of scripts that automate certain tasks, whether it's schema deploy, or pulling certain pieces of information. It sounds like you're just productizing all of that and turning it into an out-of-the-box, non-hacky, non-duct-tape solution.
```
